src/ripple/app/paths/impl/PaySteps.cpp

```diff
@@ -0,0 +1,533 @@
+                result.push_back (std::move (msr.second));
+                Currency dummy;
+                impliedPE.emplace (STPathElement::typeAccount,
+                    curIssue.account, dummy, curIssue.account);
+                cur = &*impliedPE;
+            }
+        }
+        else if (isOffer (cur) && isAccount (next))
+        {
+            if (curIssue.account != next->getAccountID () &&
+                !isXRP (next->getAccountID ()))
+            {
+                JLOG (j.trace) << "Inserting implied account after offer";
+                auto msr = make_DirectStepI (ctx(), curIssue.account,
+                    next->getAccountID (), curIssue.currency);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            continue;
+        }
+
+        if (!isOffer (next) &&
+            hasCurrency(next) && next->getCurrency () != curIssue.currency)
+        {
+            auto const& nextCurrency = next->getCurrency ();
+            auto const& nextIssuer =
+                hasIssuer (next) ? next->getIssuerID () : curIssue.account;
+
+            if (isXRP (curIssue.currency))
+            {
+                JLOG (j.trace) << "Inserting implied XI offer";
+                auto msr = make_BookStepXI (
+                    ctx(), {nextCurrency, nextIssuer});
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else if (isXRP (nextCurrency))
+            {
+                JLOG (j.trace) << "Inserting implied IX offer";
+                auto msr = make_BookStepIX (ctx(), curIssue);
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            else
+            {
+                JLOG (j.trace) << "Inserting implied II offer";
+                auto msr = make_BookStepII (
+                    ctx(), curIssue, {nextCurrency, nextIssuer});
+                if (msr.first != tesSUCCESS)
+                    return {msr.first, Strand{}};
+                result.push_back (std::move (msr.second));
+            }
+            impliedPE.emplace (
+                boost::none, nextCurrency, nextIssuer);
+            cur = &*impliedPE;
+            curIssue.currency = nextCurrency;
+            curIssue.account = nextIssuer;
+        }
+
+        auto s =
+            toStep (ctx (/*isLast*/ i == pes.size () - 2), cur, next, curIssue);
+        if (s.first == tesSUCCESS)
+            result.emplace_back (std::move (s.second));
+        else
+        {
+            JLOG (j.warning) << "toStep failed";
+            return {s.first, Strand{}};
+        }
+    }
+
+    return {tesSUCCESS, std::move (result)};
+}
+
+std::pair<TER, std::vector<Strand>>
+toStrands (
```

Comment:
So many lambdas (and lambdas embedded in lambdas) between this function and the one immediately before it that it's hard to follow. I'm wondering if we could get slightly more readable code by abstracting those two functions to a class, with private members and a single public interface.